### PR TITLE
Fixes a timing bug in Cesium map which resulted in some DataSources not getting completely removed.

### LINKED ISSUE #6344: Datasources might not be added / removed correctly from Cesium and Leaflet maps in some rare cases
After Cesiums' switch to real `Promise` api from `when.js`, the `DataSourceCollection.add()` api is fully asynchronous all the time. Previously, when passing a concrete `DataSource` as argument, the `add()` used to happen instantly without any delay. But after the switch, `add()` is [always deferred](https://github.com/CesiumGS/cesium/blob/958e74bdab1e76767ff2ab88aef0b2b05f3c956e/Source/DataSources/DataSourceCollection.js#L90). 

This has some consequence to our code for adding datasources to cesium/leaflet maps. Our implementation (incorrectly) depended on this "instant" add effect to work correctly.

- The subsequent call to [raiseToTop()](https://github.com/TerriaJS/terriajs/blob/bd6198e66b9cf4314a95464acba8ab5d40516ef8/lib/Models/Cesium.ts#L578) might silently delete a datasource (because calling `raiseToTop` on an item that does not exist has an undefined behavior)
- Datasources might get addded multiple times to the `DataSourceCollection`

This is rare and difficult to repro by adding items from catalog, but might happen when programatically adding data sources.

### PR DESCRIPTION
**Pull request by na9da**
### What this PR does

Partially fixes https://github.com/TerriaJS/terriajs/issues/6344 (for Cesium map).

I tried the same fix for Leaflet and it only seems to partially fix it, which hints at an additional bug somewhere in LeafletDataSourceDisplay. I can re-visit Leaflet later when I have time.

### Test me

It is slightly hard to replicate since this is a timing bug. But you can try the following a few times to get it.
1. Open the [main branch CI](ci.terria.io/main/) - make sure you are in 3d mode
2. Add all of these items under the `CSV` group by pressing the `+` button
image
3. Once all the items are loaded, click the `Remove all` workbench button.
4. You should sometimes see some left over points on the map, like below:
    
image
5. Now repeat the same a few times for [this branch](http://ci.terria.io/fix-cesium-ds-remove/).
6. The bug should hopefully be gone.

### Checklist

- [ ] ~There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)~ There already exists tests for this, but it is not triggered because it is a timing bug.
- [ ] I've updated relevant documentation in `doc/`.
- [ ] I've updated CHANGES.md with what I changed.
- [x] I've provided instructions in the PR description on how to test this PR.

### CODE REVIEW

File: lib/Models/Cesium.ts [hunk 1]
```diff
@@ -627,32 +627,59 @@ export default class Cesium extends GlobeOrMap {
...
+    // Ensure stacking order matches order in allDataSources - first item appears on top.
+    runInAction(() =>
+      availableDataSources.forEach((d) => dataSources.raiseToTop(d))
+    );
+  }
+
   private observeModelLayer() {
     let prevMapItems: MapItem[] = [];
-    return autorun(() => {
+    return autorun(async () => {
```
Comment (tephenavies) [inline lib/Models/Cesium.ts]: Please remove this async annotation

**Review by tephenavies — COMMENTED** [review]
This is a good improvement. I think this will work in almost all cases, but I think it's possible to have edge cases if we get two calls of syncDataSourceCollection running at once.

This may be overkill, but I think it's possible to ensure that the datasource syncing doesn't run twice simultaneously by:
* Splitting the `autorun` into sync parts and a separate reaction for the datasource syncing
* ~Use a `scheduler` to ensure the reaction function doesn't run until the last instance finishes~ Note: below I've basically recreated the `scheduler` feature of `reaction`, because I think the intent is more clear that way
* Maybe even expose the all-datasources-added promise as a public property of TerriaViewer (or GlobeOrMap)?

Some code to show the above:
```
private observeModelLayer() {
  // Closure variables
  let prevMapItems = [];
  // Instead of closure variable, this variable might be on TerriaViewer 
  let allDatasourcesAdded = Promise.resolve();
  // Reaction clean up code omitted
  autorun(() => {
    // Existing autorun that deals with Primitives and ImageryLayers
  });
  reaction(() => this.availableDataSources, () => {
    allDatasourcesAdded = allDatasourcesAdded.then(
      () => this.syncDataSourceCollection(this.availableDataSources, this.dataSources)
    );
  });
  return cleanupFunc;
}
```

It's possible that calls of the syncDSC function will be delayed so they end up running multiple times, but I think it won't be able to run simultaneously to itself (which I think could cause problems).

**Comment by tephenavies** [comment]
What do you think @na9da?

**Comment by tephenavies** [comment]
Actually I think my suggestion has a problem that the allDatasourcesAdded promise is not set synchronously when a change to mapItems is made, so you'd still have to do e.g.:

```
// Manipulate workbench
workbench.add(item)
// Wait 0ms for reaction inside Cesium.ts to run
await Promise.resolve();
// Wait potentially for an async datasource
await terriaViewer.allDatasourcesAdded;
// Do something with cesium.dataSources (e.g. check for a dataset in a spec)
```

**Comment by na9da** [comment]
Nice catch @steve9164. For a moment I forgot the root cause of the bug which is that the `.contains()` check and the[ final push() to the collection](https://github.com/TerriaJS/cesium/blob/160d6f108c8bf606fd608df04d375fec5c3f7cb7/Source/DataSources/DataSourceCollection.js#L90-L94) does not happen synchronously which leaves a gap for concurrent calls to slip through and do multiple inserts. 

If I understood correctly, in your solution, there is still a chance that different concurrent runs get the same closure value for the promise and so fire concurrently when the previous promise finishes? 

Another possible solution is to maintain a `Set<DataSource>` of active "syncs" as a guard against double inserts.
```
class Cesium {
   // 
   private activeDataSourceSyncs: Set<DataSource> = new Set();

   syncDataSourceCollection(...) {
        ...
        for (let ds of availableDataSources) {
           if (!dataSources.contains(ds) && !this.activeDataSourceSyncs.has(ds)) {
               try {
                 activeDataSourceSyncs.add(ds);
                 await dataSources.add(ds);
               } finally {
                 activeDataSourceSyncs.remove(ds);
               }
           }
        }
   }
}
```

Longer term solution would be to contribute an `addIfDoesNotContain()` or `addSync()` method upstream as it looks like this could be a common source of bugs. Otherwise, we could create a child class of `DataSourceCollection` to add them.

However, I would prefer something similar to your solution as it could ensure that the whole syncing thing happens only 1 at at time.

**Comment by na9da** [comment]
Another solution:

- Keep a sync count with a maximum value of 2
- Inisde the sync() method, skip execution if the current sync count > 0 (meaning a sync is already running)
- When the current active run finishes, it decrements the sync count and re-runs sync() if the sync count is still `> 0` (with setTimeout so that we don't enter a spin loop)

The max value 2 is because even if we get 10s of re-runs we only need to sync with the latest `availableDataSources`.

**Comment by na9da** [comment]
Hey @steve9164, I have implemented my 2nd solution of using a flag/count to skip syncing when another sync is already running. 

I think I have also figured out the problem with syncing for Leaflet. The `raiseToTop(ds)` call silently deletes the last item in the collection if `ds` does not exist. For Leaflet, when we set `show=false` for a workbench item, we are actually removing the DS from the collection. After that, when we do `raiseToTop(ds)` if we don't check for `contains(ds)` first, it will not only add `ds` but also remove the last element from the collection!

The latest commit should also hopefully fix #5335

**Comment by tephenavies** [comment]
@na9da 
> If I understood correctly, in your solution, there is still a chance that different concurrent runs get the same closure value for the promise and so fire concurrently when the previous promise finishes?

I don't think that's true. In my snippet when `allDatasourcesAdded` is used, it's immediately updated with a new promise that won't resolve until the new async task is complete. If the reaction fires twice, execution 1 will update the promise and execution 2 will immediately attach to the new promise updated by 1 (which resolves after all async work in 1 is done).

**Comment by na9da** [comment]
@steve9164, I have updated Cesium & Leaflet to use your solution and tested it. It looks more elegant now.

File: lib/Models/Leaflet.ts [hunk 2]
```diff
@@ -403,37 +450,55 @@ export default class Leaflet extends GlobeOrMap {
...
+      if (!dataSources.contains(ds) && ds.show) {
+        await dataSources.add(ds);
+      }
+    }
 
-      // Ensure stacking order matches order in allDataSources - first item appears on top.
-      allDataSources.forEach((d) => dataSources.raiseToTop(d));
-    });
+    await new Promise((resolve) =>
+      window.setTimeout(resolve, Math.random() * 1000)
```
Comment (tephenavies) [inline lib/Models/Leaflet.ts]: Is this a remnant of testing?

Comment (na9da) [reply lib/Models/Leaflet.ts]: Ooops! removed.

**Review by tephenavies — APPROVED** [review]

**Pull request merged**